tuya.cpp

```diff
@@ -172,19 +176,90 @@ void DeRestPluginPrivate::handleTuyaClusterIndication(const deCONZ::ApsDataIndic
         DBG_Printf(DBG_INFO, "Tuya debug 4 : Address 0x%016llX Payload %s\n" , ind.srcAddress().ext(), qPrintable(zclFrame.payload().toHex()));
         DBG_Printf(DBG_INFO, "Tuya debug 5 : Status: %d Transid: %d Dp: %d (0x%02X,0x%02X) Fn: %d Data %ld\n", status , transid , dp , dp_type, dp_identifier, fn , data);
         
-        //Sensor and light use same cluster, so need to make a choice for device that have both
-        if ((sensorNode) && (lightNode))
+        if (length > 4) //schedule command
+        {
+            
+            // Monday = 64, Tuesday = 32, Wednesday = 16, Thursday = 8, Friday = 4, Saturday = 2, Sunday = 1
+            // If you want your schedule to run only on workdays, the value would be W124. (64+32+16+8+4 = 124)
+            // The API specifies 3 numbers, so a schedule that runs on Monday would be W064.
+            // Workday = W124
+            // Not working day = W003
+            // Saturday = W002
+            // Sunday = W001
+            // All days = W127
+            
+            QString transitions = QString("");
+            
+            length = length / 3;
+            
+            if (zclFrame.payload().size() < (( length * 3) + 6) )
+            {
+                DBG_Printf(DBG_INFO, "Tuya : Schedule data error\n");
+                return;
+            }
+            
+            quint8 hour;
+            quint8 minut;
+            quint8 heatSetpoint;
+            
+            quint8 part = 1;
+            QList<int> listday;
+            
+            if (dp == 0x0070) //work days (6)
+            {
+                listday << 124;
+            }
+            else if (dp == 0x00071) // holiday = Not working day (6)
+            {
+                listday << 003;
+            }
+            else if (dp == 0x0065) // Moe thermostat W124 (4) + W002 (4) + W001 (4)
+            {
+                part = length / 3;
+                listday << 124 << 2 << 1;
+            }
+            
+            for (; part > 0; part--)
+            {
+                for (; length > 0; length--)
+                {
+                    stream >> hour;
+                    stream >> minut;
+                    stream >> heatSetpoint;
+
+                    transitions += QString("T%1:%2|%3")
+                        .arg(hour, 2, 10, QChar('0'))
+                        .arg(minut, 2, 10, QChar('0'))
+                        .arg(heatSetpoint);
+                }
+            }
+            
+            updateThermostatSchedule(sensorNode, listday.at(part - 1), transitions);
```

Comment:
When I read this correctly `part` is always 0 at this point due the for loop and part - 1 becomes `0 - 1 == 255`.
`listday` isn't checked to be non empty, which happens when a dp value isn't handled above.

Therefore as simple guard I'd propose:

```
if (part > 0 && listday.size() >= static_cast<int>(part))
{
    updateThermostatSchedule(sensorNode, listday.at(part - 1), transitions);
}
```

Reply:
There is 
>quint8 part = 1;

So this value can't be 0 ?
But yes the guard is usefull, and I have see another problem this line is not a good place. Not enought time to test it with users, will be for next month.

Reply:
If it becomes 0 the code would crash since the access to `listday` at index 255 (due the unsigned integer wrap). The guard would prevent this case.

Reply:
Done